File: test/python-tests/test_assertions.cpp
#include <iostream>
#include <string>
#include <vector>
#include <optional>
#include <unordered_map>
#include <tuple>
#include <utility>
#include <type_traits>
#include <functional>
#include <sstream>
#include <iomanip>
#include <cmath>
#include <algorithm>
#include <typeinfo>
#include <cstring>

using namespace std;

// helpers
template<typename T> static inline std::ostream& operator<<(std::ostream& os, const std::vector<T>& v);
template<typename T> static inline std::string to_str(const T& v){ std::ostringstream os; os<<v; return os.str(); }
static inline std::string to_str(const std::string& s){ return s; }
static inline std::string to_str(const char* s){ return std::string(s); }
static inline std::string to_str(double v){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(1)<<v; return os.str(); }
static inline std::string to_str(float v){ std::ostringstream os; os.setf(std::ios::fixed); os<<std::setprecision(1)<<v; return os.str(); }
template<typename T> static inline int to_int(const T& v){ if constexpr (std::is_convertible_v<T,std::string>) return std::stoi(v); else return static_cast<int>(v); }
template<typename T> static inline float to_float(const T& v){ if constexpr (std::is_convertible_v<T,std::string>) return std::stof(v); else return static_cast<float>(v); }
template<typename C> static inline int sum(const C& c){ int s=0; for(const auto& e: c) s += e; return s; }
template<typename T> static inline int sum(std::initializer_list<T> c){ int s=0; for(const auto& e: c) s += e; return s; }
template<typename T> static inline size_t len(const T& c){ return c.size(); }
static inline size_t len(const char* s){ return std::char_traits<char>::length(s); }
template<typename K, typename V> static inline std::string to_str(const std::unordered_map<K,V>&){ return std::string("{...}"); }
// minimal ABC base to support 'class X : public ABC'
struct ABC {};
// minimal math namespace replacement for 'math.pi'
struct __fluxus_math { static constexpr double pi = 3.141592653589793; static inline double sqrt(double x){ return std::sqrt(x); } } math;
// minimal sys module stub
namespace sys { static std::vector<std::string> argv; }
namespace asyncio { template<typename F> static inline void run(F f){ f(); } }

// vector to string conversion for printing
template<typename T>
static inline std::string vec_to_str(const std::vector<T>& v) {
    std::ostringstream os;
    os << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i > 0) os << ", ";
        if constexpr (std::is_same_v<T, std::string>) {
            os << "'" << v[i] << "'";
        } else {
            os << v[i];
        }
    }
    os << "]";
    return os.str();
}

// nested vector to string
template<typename T>
static inline std::string vec_to_str(const std::vector<std::vector<T>>& v) {
    std::ostringstream os;
    os << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i > 0) os << ", ";
        os << vec_to_str(v[i]);
    }
    os << "]";
    return os.str();
}

// stream operator for vectors
template<typename T>
static inline std::ostream& operator<<(std::ostream& os, const std::vector<T>& v) {
    os << vec_to_str(v);
    return os;
}
template<typename K, typename V> static inline std::ostream& operator<<(std::ostream& os, const std::unordered_map<K,V>&){ os << "{...}"; return os; }
template<typename K, typename V> static inline auto py_items(const std::unordered_map<K,V>& m){ std::vector<std::pair<K,V>> v; v.reserve(m.size()); for(const auto& kv: m){ v.emplace_back(kv.first, kv.second);} return v; }


class Rectangle {
    public:
        Rectangle(int width, int height) {
            this->width = width;
            this->height = height;
        }
        int area() {
    auto result = this->width * this->height;
    assert result > 0, "Area must be positive";
    return result;
}


    public:
        int width = 0;
        int height = 0;
};


int divide(int a, int b) {
    auto assert b ! = 0, "Divisor cannot be zero";
    return a / b;
}


bool validate_person(std::vector<int> name, int age) {
    assert true;
    assert len(name) > 0, "Name cannot be empty";
    assert true;
    auto assert age > = 0, "Age cannot be negative";
    auto assert age < = 150, "Age seems unrealistic";
    return true;
}


bool is_prime(int n) {
    if (n < 2) {
        return false;
    }
    for (int i = 2; i < int(n ** 0.5) + 1; ++i) {
        if (n % i == 0) {
            return false;
        }
    }
    return true;
}


std::string test_with_try_except() {
    bool __fluxus_exc=false;
    {
        assert false, "This assertion should fail";
    }
    if (__fluxus_exc) { auto e = 0;
        std::cout << to_str("Caught assertion error: ") << to_str(e) << std::endl;
        return "Handled";
    }
    return "Not reached";
}


int main() {
    bool __fluxus_exc=false;
    std::cout << "=== Basic Assert ===" << std::endl;
    auto x = 5;
    auto assert x = = 5;
    std::cout << "Assert x == 5 passed" << std::endl;
    assert x > 0;
    std::cout << "Assert x > 0 passed" << std::endl;
    std::cout << "\n=== Assert with Message ===" << std::endl;
    auto age = 25;
    auto assert age > = 18, "Must be at least 18 years old";
    std::cout << "Age assertion passed" << std::endl;
    std::cout << "\n=== Complex Expression Assert ===" << std::endl;
    auto numbers = std::vector<decltype(1)>{1, 2, 3, 4, 5};
    auto assert len(numbers) = = 5, f"Expected 5 elements, got {len(numbers)}";
    std::cout << "Length assertion passed" << std::endl;
    auto assert sum(numbers) = = 15, "Sum should be 15";
    std::cout << "Sum assertion passed" << std::endl;
    std::cout << "\n=== Type Checking Assert ===" << std::endl;
    auto value = "hello";
    assert true;
    std::cout << "Type assertion passed" << std::endl;
    auto number = 42;
    assert true;
    std::cout << "Number type assertion passed" << std::endl;
    std::cout << "\n=== Assert in Functions ===" << std::endl;
    auto result = divide(10, 2);
    std::cout << to_str("divide(10, 2) = ") << to_str(result) << std::endl;
    {
    }
    if (__fluxus_exc) { auto e = 0;
    }
    std::cout << "\n=== Boolean Expression Assert ===" << std::endl;
    auto is_valid = true;
    assert is_valid, "Validation failed";
    std::cout << "Validation assertion passed" << std::endl;
    auto is_empty = false;
    assert not is_empty, "Should not be empty";
    std::cout << "Empty check assertion passed" << std::endl;
    std::cout << "\n=== Multiple Assertions ===" << std::endl;
    auto result = validate_person("Alice", 25);
    std::cout << "All person validations passed" << std::endl;
    std::cout << "\n=== Assert with Membership ===" << std::endl;
    auto valid_colors = std::vector<std::string>{'red', 'green', 'blue'};
    auto color = 'red';
    assert color in valid_colors, f"Color {color} not in valid colors";
    std::cout << "Color membership assertion passed" << std::endl;
    std::cout << "\n=== Assert with Range ===" << std::endl;
    auto temperature = 25;
    auto assert 0 < = temperature <= 100, "Temperature out of range";
    std::cout << "Temperature range assertion passed" << std::endl;
    std::cout << "\n=== Assert with List Operations ===" << std::endl;
    auto items = std::vector<decltype(1)>{1, 2, 3, 4, 5};
    0;
    std::cout << "All positive assertion passed" << std::endl;
    auto values = std::vector<decltype(2)>{2, 4, 6, 8};
    auto assert all(x % 2 = 0;
    std::cout << "All even assertion passed" << std::endl;
    std::cout << "\n=== Assert with any ===" << std::endl;
    auto numbers = std::vector<decltype(1)>{1, 3, 5, 7, 8};
    auto assert any(x % 2 = 0;
    std::cout << "Any even assertion passed" << std::endl;
    std::cout << "\n=== Assert with String Operations ===" << std::endl;
    auto text = "Hello, World!";
    assert text.startswith("Hello"), "Text should start with Hello";
    std::cout << "String start assertion passed" << std::endl;
    assert "World" in text, "Text should contain World";
    std::cout << "String contains assertion passed" << std::endl;
    std::cout << "\n=== Assert with Dictionary ===" << std::endl;
    auto person = std::unordered_map<std::string, std::string>{{"name", 'Alice'}, {"age", to_str(25)}};
    assert 'name' in person, "Person must have a name";
    assert 'age' in person, "Person must have an age";
    std::cout << "Dictionary key assertions passed" << std::endl;
    auto assert person['age'] > = 18, "Person must be an adult";
    std::cout << "Dictionary value assertion passed" << std::endl;
    std::cout << "\n=== Assert with Custom Conditions ===" << std::endl;
    auto num = 7;
    assert is_prime(num), f"{num} should be prime";
    std::cout << to_str("Prime assertion for ") << to_str(num) << to_str(" passed") << std::endl;
    std::cout << "\n=== Assert with Class Attributes ===" << std::endl;
    auto rect = Rectangle(5, 10);
    auto area = rect.area();
    std::cout << to_str("Rectangle area: ") << to_str(area) << std::endl;
    std::cout << "Class assertion tests passed" << std::endl;
    std::cout << "\n=== Assert with Comparison Chains ===" << std::endl;
    auto x = 5;
    assert 0 < x < 10, "x must be between 0 and 10";
    std::cout << "Comparison chain assertion passed" << std::endl;
    auto values = std::vector<decltype(1)>{1, 2, 3, 4, 5};
    for (auto v : values) {
    }
    std::cout << "All values in range assertion passed" << std::endl;
    std::cout << "\n=== Testing Assertion Failure Handling ===" << std::endl;
    auto result = test_with_try_except();
    std::cout << to_str("Result: ") << to_str(result) << std::endl;
    std::cout << "\n=== All assertion tests completed ===" << std::endl;
    return 0;
}

File: test/python-tests/test_assertions.py
# Test assert statements and assertion features

# Basic assert
print("=== Basic Assert ===")
x = 5
assert x == 5
print("Assert x == 5 passed")

assert x > 0
print("Assert x > 0 passed")

# Assert with message
print("\n=== Assert with Message ===")
age = 25
assert age >= 18, "Must be at least 18 years old"
print("Age assertion passed")

# Assert with complex expression
print("\n=== Complex Expression Assert ===")
numbers = [1, 2, 3, 4, 5]
assert len(numbers) == 5, f"Expected 5 elements, got {len(numbers)}"
print("Length assertion passed")

assert sum(numbers) == 15, "Sum should be 15"
print("Sum assertion passed")

# Assert with type checking
print("\n=== Type Checking Assert ===")
value = "hello"
assert isinstance(value, str), "Value must be a string"
print("Type assertion passed")

number = 42
assert isinstance(number, int), "Number must be an integer"
print("Number type assertion passed")

# Assert in functions
print("\n=== Assert in Functions ===")
def divide(a, b):
    assert b != 0, "Divisor cannot be zero"
    return a / b

result = divide(10, 2)
print(f"divide(10, 2) = {result}")

try:
    result = divide(10, 0)
except AssertionError as e:
    print(f"Caught expected assertion: {e}")

# Assert with boolean expressions
print("\n=== Boolean Expression Assert ===")
is_valid = True
assert is_valid, "Validation failed"
print("Validation assertion passed")

is_empty = False
assert not is_empty, "Should not be empty"
print("Empty check assertion passed")

# Multiple assertions
print("\n=== Multiple Assertions ===")
def validate_person(name, age):
    assert isinstance(name, str), "Name must be a string"
    assert len(name) > 0, "Name cannot be empty"
    assert isinstance(age, int), "Age must be an integer"
    assert age >= 0, "Age cannot be negative"
    assert age <= 150, "Age seems unrealistic"
    return True

result = validate_person("Alice", 25)
print("All person validations passed")

# Assert with membership
print("\n=== Assert with Membership ===")
valid_colors = ['red', 'green', 'blue']
color = 'red'
assert color in valid_colors, f"Color {color} not in valid colors"
print("Color membership assertion passed")

# Assert with range
print("\n=== Assert with Range ===")
temperature = 25
assert 0 <= temperature <= 100, "Temperature out of range"
print("Temperature range assertion passed")

# Assert with list operations
print("\n=== Assert with List Operations ===")
items = [1, 2, 3, 4, 5]
assert all(x > 0 for x in items), "All items must be positive"
print("All positive assertion passed")

values = [2, 4, 6, 8]
assert all(x % 2 == 0 for x in values), "All values must be even"
print("All even assertion passed")

# Assert with any
print("\n=== Assert with any ===")
numbers = [1, 3, 5, 7, 8]
assert any(x % 2 == 0 for x in numbers), "At least one even number required"
print("Any even assertion passed")

# Assert with string operations
print("\n=== Assert with String Operations ===")
text = "Hello, World!"
assert text.startswith("Hello"), "Text should start with Hello"
print("String start assertion passed")

assert "World" in text, "Text should contain World"
print("String contains assertion passed")

# Assert with dictionary
print("\n=== Assert with Dictionary ===")
person = {'name': 'Alice', 'age': 25}
assert 'name' in person, "Person must have a name"
assert 'age' in person, "Person must have an age"
print("Dictionary key assertions passed")

assert person['age'] >= 18, "Person must be an adult"
print("Dictionary value assertion passed")

# Assert with custom conditions
print("\n=== Assert with Custom Conditions ===")
def is_prime(n):
    if n < 2:
        return False
    for i in range(2, int(n ** 0.5) + 1):
        if n % i == 0:
            return False
    return True

num = 7
assert is_prime(num), f"{num} should be prime"
print(f"Prime assertion for {num} passed")

# Assert with class attributes
print("\n=== Assert with Class Attributes ===")
class Rectangle:
    def __init__(self, width, height):
        assert width > 0, "Width must be positive"
        assert height > 0, "Height must be positive"
        self.width = width
        self.height = height
    
    def area(self):
        result = self.width * self.height
        assert result > 0, "Area must be positive"
        return result

rect = Rectangle(5, 10)
area = rect.area()
print(f"Rectangle area: {area}")
print("Class assertion tests passed")

# Assert with comparison chains
print("\n=== Assert with Comparison Chains ===")
x = 5
assert 0 < x < 10, "x must be between 0 and 10"
print("Comparison chain assertion passed")

values = [1, 2, 3, 4, 5]
for v in values:
    assert 0 < v <= 5, f"Value {v} out of expected range"
print("All values in range assertion passed")

# Testing assertion failure handling
print("\n=== Testing Assertion Failure Handling ===")
def test_with_try_except():
    try:
        assert False, "This assertion should fail"
    except AssertionError as e:
        print(f"Caught assertion error: {e}")
        return "Handled"
    return "Not reached"

result = test_with_try_except()
print(f"Result: {result}")

print("\n=== All assertion tests completed ===")
